From e6370bee840b2c70df959402d7feeb09d3215d8e Mon Sep 17 00:00:00 2001
From: Colin Guthrie <colin@mageia.org>
Date: Mon, 13 Jan 2014 11:06:35 +0000
Subject: [PATCH] systemctl: Do not attempt native calls for enable/disable if
 sysvinit handleds all units.

If you have sysvinit compat enabled, it might deal with all the passed
in unit names leaving only a null termitated strv structure.

This cannot be iterated (sensibly) and thus we should just bail out
and not bother calling various native methods.

This should prevent a bogus warning regarding the lack of [Install]
sections in systemd units when enabling a sysvinit unit.
---
 src/systemctl/systemctl.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/src/systemctl/systemctl.c b/src/systemctl/systemctl.c
index a9d948e..9342793 100644
--- a/src/systemctl/systemctl.c
+++ b/src/systemctl/systemctl.c
@@ -5292,6 +5292,11 @@ static int enable_unit(sd_bus *bus, char **args) {
                 return r;
 
         if (!bus || avoid_bus()) {
+                /* If the sysvinit compat calls have handled everything we may
+                 * have none left to deal with natively... */
+                if (!names[0])
+                        goto finish;
+
                 if (streq(verb, "enable")) {
                         r = unit_file_enable(arg_scope, arg_runtime, arg_root, names, arg_force, &changes, &n_changes);
                         carries_install_info = r;
@@ -5330,6 +5335,11 @@ static int enable_unit(sd_bus *bus, char **args) {
 
                 polkit_agent_open_if_enabled();
 
+                /* If the sysvinit compat calls have handled everything we may
+                 * have none left to deal with natively... */
+                if (!names[0])
+                        goto reload;
+
                 if (streq(verb, "enable")) {
                         method = "EnableUnitFiles";
                         expect_carries_install_info = true;
@@ -5404,6 +5414,7 @@ static int enable_unit(sd_bus *bus, char **args) {
                 if (r < 0)
                         return r;
 
+reload:
                 /* Try to reload if enabled */
                 if (!arg_no_reload)
                         r = daemon_reload(bus, args);
-- 
2.3.2

